src/Core/PluginRegistry.php

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Plugin Registry
+ */
+class PluginRegistry extends ObjectRegistry implements PluginRegistryInterface
+{
+
+    /**
+     * Should resolve the classname for a given object type.
+     *
+     * @param string $class The class to resolve.
+     * @return string|bool The resolved name or false for failure.
+     */
+    protected function _resolveClassName($class)
+    {
+        if (class_exists($class)) {
+            return $class;
+        }
+
+        list($plugin, $name) = pluginSplit($class, true);
+
+        if (empty($plugin)) {
+            $class = $name . '\\' . 'Plugin';
+        } else {
+            $class = $plugin . $name;
+        }
+
+        return $class;
+    }
+
+    /**
+     * Throw an exception when the requested object name is missing.
+     *
+     * @param string $class The class that is missing.
+     * @param string $plugin The plugin $class is missing from.
+     * @return void
+     * @throws \Exception
+     */
+    protected function _throwMissingClassError($class, $plugin)
+    {
+        new RuntimeException(sprintf(
+            'Plugin class `%s` not found.',
+            $class
+        ));
+    }
+
+    /**
+     * Create an instance of a given classname.
+     * This method should construct and do any other initialization logic
+     * required.
+     *
+     * @param string $class The class to build.
+     * @param string $alias The alias of the object.
+     * @param array $config The Configuration settings for construction
+     * @return mixed
+     */
+    protected function _create($class, $alias, $config)
+    {
+        $instance = new $class($config);
+
+        if ((!$instance instanceof HttpApplicationInterface
+            && !$instance instanceof ConsoleApplicationInterface)
+            || !$instance instanceof PluginInterface) {
+            throw new RuntimeException(sprintf(
+                '`%s` is not a valid plugin object. It\'s not implementing `%s` or `%s`',
+                get_class($instance)),
+                HttpApplicationInterface::class,
+                ConsoleApplicationInterface::class
+            );
+        }
+
+        return $instance;
+    }
+
+    /**
+     * Load all the application configuration and bootstrap logic.
+     * Override this method to add additional bootstrap logic for your application.
+     *
+     * @return void
+     */
+    public function bootstrap()
+    {
+        foreach ($this as $plugin) {
+            if ($plugin->isBootstrapEnabled()) {
+                $plugin->bootstrap();
+            }
+        }
+    }
+
+    /**
+     * Define the console commands for an application.
+     *
+     * @param \Cake\Console\CommandCollection $commands The CommandCollection to add commands into.
+     * @return \Cake\Console\CommandCollection The updated collection.
+     */
+    public function console($commands)
+    {
+        foreach ($this as $plugin) {
+            if ($plugin instanceof ConsoleApplicationInterface) {
```

Comment:
Since you have `PluginApp implements ConsoleApplicationInterface` doesn't this check become redundant? IMO the base class should not implement `ConsoleApplicationInterface`, only plugins which provide console commands should implement that interface.

Reply:
Correct @robertpustulka

Reply:
Gotcha, my bad.